Feature Safe Places

Roleen20 opened the pull request:
Integration with safe places

Roleen20 committed:
Implementation for the safe places feature.
Added broker and exception handling.
Added the service to retrieve the last two weeks data of a user.

Roleen20 committed:
Registering the http client

Roleen20 committed:
Merge branch 'develop' into feature/cov335

Roleen20 committed:
Bug fix for the location to pull through. Some validation checks

Roleen20 committed:
Commenting entry out for if the data is merge, this will allow dev to not break

Roleen20 committed:
Integration with the organisation when a user is checking in and is postive

Roleen20 committed:
Undo the comments

deltaDirk commented on V1/Controllers/OrganisationController.cs:
```diff
@@ -57,11 +57,11 @@ public async Task<IActionResult> GetOrganisation(string id)
 
-        [HttpPost("organisations/{id}/check_in")]
-        public async Task<IActionResult> CheckIn(string id, [FromBody] UpdateCountRequest payload)
+        [HttpPost("organisations/{id}/check_in/{isPositive}")]
```
This can not happen. We need to determine whether the person checking in is positive. This can't be just passed via the app. Remember, the only thing the app has context of here is the user's QR code ie the walletId.

deltaDirk commented on V1/Controllers/OrganisationController.cs:
```diff
@@ -78,10 +78,10 @@ public async Task<IActionResult> AccessDenied(string id, [FromBody] UpdateCountR
 
-        [HttpPost("organisations/{organisationId}/mobile_check_in")]
-        public async Task<IActionResult> MobileCheckIn(string organisationId, [FromBody] MobileUpdateCountRequest payload)
+        [HttpPost("organisations/{organisationId}/mobile_check_in/{isPositive}")]
```
It is impossible to know whether the person is tested positive here. We only have their number available.

deltaDirk commented on V1/Repositories/OrganisationAccessLogRepository.cs:
```diff
@@ -36,5 +36,13 @@ public async Task<List<OrganisationAccessLog>> GetListByWalletIds(List<Guid> wal
+
+        public async Task<List<OrganisationAccessLog>> GetLogsForLastTwoWeeks(Guid walletId, DateTime testedAt)
+        {
+          return await _dbSet.Where(oal => oal.Wallet.Id.ToString().Equals(walletId.ToString()))
```
Can just use Equals method here. Don't have to convert to strings first.

Roleen20 replied:
It broke for some reason when i did that

deltaDirk commented on V1/Repositories/OrganisationAccessLogRepository.cs:
```diff
@@ -36,5 +36,13 @@ public async Task<List<OrganisationAccessLog>> GetListByWalletIds(List<Guid> wal
+        {
+          return await _dbSet.Where(oal => oal.Wallet.Id.ToString().Equals(walletId.ToString()))
+                .Where(oal => oal.CreatedAt >= testedAt.AddDays(-14) && oal.ScanType == ScanType.CheckIn)
+                .Include(oal => oal.Wallet)
```
We already have the walletId here, we don't need to include the wallet associated with every log record

deltaDirk commented on V1/Services/OrganisationService.cs:
```diff
@@ -86,7 +89,7 @@ public async Task<Response> GetAsync(string id)
         }
 
-        public async Task<Response> UpdateCountAsync(string id, UpdateCountRequest payload, ScanType scanType, bool mobile = false)
+        public async Task<Response> UpdateCountAsync(string id, UpdateCountRequest payload, ScanType scanType, bool mobile = false, bool isPositive = false)
```
Here is where we need to determine whether the person is infected. This can be done by just checking their test results.

deltaDirk commented on V1/Services/StaySafeService.cs:
```diff
@@ -0,0 +1,62 @@
+                        trails.Add(new Trail
+                        {
+                            Time = log.CreatedAt.ToString(),
+                            Latitude = decimal.ToDouble(log.Latitude),
```
Maybe just round this to 2 decimals. Also is it necessary to convert this to a double?

Roleen20 replied:
The safe places API accepts double. If i round the value it will still not be accepted.

deltaDirk commented on V1/Services/StaySafeService.cs:
```diff
@@ -0,0 +1,62 @@
+                {
+                    var request = new RedactedRequest
+                    {
+                        Identifier = Guid.NewGuid().ToString(),
```
Do we maybe not need to record this somehow?

Roleen20 replied:
I don't know, I thought about adding the wallet Id but do we want to know who was were or do we just want to know were a sick person was?

deltaDirk commented on V1/Services/TestResultService.cs:
```diff
@@ -73,6 +76,9 @@ public async Task AddTestResult(TestResultRequest testResultRequest)
             await _walletTestResultRepository.SaveAsync();
+
+            if (testResultRequest.ResultStatus == ResultStatus.Positive)
+                BackgroundJob.Enqueue(() => _staySafeService.CaptureData(wallet.Id, testResultRequest.TestedAt));
```
Just add this to the other AddTestResult function. Or maybe just get rid of the second and adjust the call from the OTP side

Roleen20 replied:
I don't think we are using that flow on the frontend to submit the wallet and test results in the otp at once. But ill update it

Roleen20 committed:
PR review addtions.
Making route optional
Updating the add test results, combining the two methods

Roleen20 committed:
Changing the query to not stringify it

Roleen20 committed:
commenting out the login

mr-django commented on V1/Controllers/OrganisationController.cs:
```diff
@@ -57,11 +57,11 @@ public async Task<IActionResult> GetOrganisation(string id)
 
-        [HttpPost("organisations/{id}/check_in")]
-        public async Task<IActionResult> CheckIn(string id, [FromBody] UpdateCountRequest payload)
+        [HttpPost("organisations/{id}/check_in/{isPositive?}")]
```
QUESTION: Does `isPositive` best describe the param? Would `{status}` not be a better description?

mr-django replied:
I did see @deltaDirk mention below that you want to test if the user is infected, so this comment might not useful

Roleen20 committed:
Merging from dev

Roleen20 committed:
Refatoring after the merge from dev.
Making the wallet receipt methods more extendable for the service to define the date

Roleen20 committed:
Adding in the is enabled switch for safe places to integrate or not

Roleen20 committed:
Merge from dev

Roleen20 committed:
Fixes for the safe path integration

Roleen20 committed:
Busy with refactoring

Roleen20 committed:
Update to send data if consent is given

Roleen20 committed:
Getting the maps api key

Roleen20 committed:
Only sending todays receipts to safe places when checking into an organisation

Roleen20 committed:
Fixes for safe places integration

Roleen20 committed:
Resolving merge conflicts from dev

Roleen20 committed:
Fixing startup

Roleen20 committed:
Updating git ignore file

Roleen20 committed:
Updating project file

Roleen20 committed:
Fixing the time stap to be a long not a string.
Refactored the login method
